clippy_lints/src/returns.rs

```diff
@@ -407,7 +407,7 @@ fn check_final_expr<'tcx>(
                 }
             }
 
-            if ret_span.from_expansion() {
+            if ret_span.from_expansion() || is_from_proc_macro(cx, expr) {
```

Comment:
Note: if this is the `Span::from_expansion` that is the rustc one, then in general it is the most robust check for spans from macro expansions that is possible, and I would not have expected `is_from_proc_macro` detection to *not* be already covered by `from_expansion` **if** the proc macro implements span hygiene correctly.

I suspect that some crates have proc-macros which generate spans that have call-site hygiene instead of mixed-site hygiene, causing `from_expansion` to return `false`.

But I suppose it's possible that clippy wants different trade-offs here and try to provide warnings even if the proc-macros do not emit spans with proper hygiene?

See: similar issue I ran into in implementing `unit_bindings` lint in rustc https://github.com/rust-lang/rust/pull/112380#issuecomment-1657124150.

Also see a [Rocket PR](https://github.com/rwf2/Rocket/pull/2586#issuecomment-1694224046) where I used `Span::mixed_site().located_at(span)` so the proc-macro produces a `Span` which will correctly return `true` for `Span::from_expansion` because it has distinguishing `SyntaxContext`.

Reply:
Yeah, the issue that `is_from_proc_macro` solves is crates using `quote_spanned!` with an input span like tokio does in the linked issues which makes it (to my knowledge) completely impossible to detect whether a span is from a proc macro, so this is a big hack that approximates it even further by checking if the code pointed at the span lines up with the AST.

`@Alexendoo` wrote a good comment about the proc macro situation in a zulip thread: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/report_in_external_macro.20no.20longer.20configurable.3F/near/448135182

To be clear, I'd also love to get rid of this, but afaik we're kind of stuck with this hack for now and add such checks whenever needed because proc macro tooling like `quote::quote_spanned!` makes it incredibly easy to fool clippy and we get a lot of bug reports for it because a lot of proc macros do this

Reply:
Right, that's entirely reasonable for clippy. I opened a zulip topic over for T-compiler (and T-lang) to discuss about the proc-macro span hygiene situation (because rustc lints/diagnostics also gets bitten by this quite often): https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/proc-macro.20span.20hygiene.